src/Validation/RulesGatherer.php

```diff
@@ -224,6 +225,10 @@ protected function qualifyArgumentReferences(array $rules, array $argumentPath):
         return array_map(
             static function ($rule) use ($argumentPath) {
                 if (is_object($rule)) {
+                    if ($rule instanceof WithReferenceRule) {
+                        $rule->setArgumentPath(implode('.', $argumentPath));
```

Comment:
We can simply pass the array, access to more structured data is generally better.